Prep/src/service/ConfigMergeAndOverrideSolution.java

```diff
@@ -0,0 +1,199 @@
+            put("key2", "val2");
+            put("key4", new HashMap<String, Object>(){{
+                put("key41", "val41");
+                put("key42", "val42");
+            }});
+        }};
+        Map<String, Object> defaultConfig = new HashMap<String, Object>(){{
+            put("key2", "val2_");
+            put("key3", "val3_");
+            put("key4", new HashMap<String, Object>(){{
+                put("key42", "val42_");
+                put("key43", "val43_");
+            }});
+        }};
+
+        // Prints {key1=val1, key2=val2, key3=val3_, key4={key43=val43_, key42=val42, key41=val41}}
+        Printer.print(configMapMergeAndOverride(appConfig, defaultConfig));
+
+        Printer.println();
+
+        // Prints {key1=val1, key2=val2, key4.key42=val42, key4.key43=val43_, key4.key41=val41, key3=val3_}
+        Printer.print(configMapMergeAndOverrideStretchGoal(appConfig, defaultConfig));
+    }
+
+    // Time complexity: O(n + m) = O(n)
+    // Space Complexity: O(n + m) = O(n)
+    public static Map<String, Object>  configMapMergeAndOverride(
+            Map<String, Object> appConfig,
+            Map<String, Object> defaultConfig) {
+
+        // input validation
+        if (appConfig == null) {
+            return defaultConfig;
+        } else if (defaultConfig == null) {
+            return appConfig;
+        } else {
+            final Map<String, Object> resultConfig = new HashMap<>();
+            // iterate over appConfig
+            appConfig.entrySet().stream()
+                    .forEach(appConfigEntry -> {
+                        final String key = appConfigEntry.getKey();
+                        Map.Entry<String, Object> defaultConfigEntry = new MapEntry<>(key, defaultConfig.get(key));
+
+                        Map.Entry<String, Object> mergedEntry =  configEntryMergeAndOverride(appConfigEntry, defaultConfigEntry);
+                        resultConfig.put(mergedEntry.getKey(), mergedEntry.getValue());
+                    });
+
+            // iterate over defaultConfig to add entries not present in appConfig
+            defaultConfig.entrySet().stream()
+                    .forEach(defaultConfigEntry -> {
+                        if (appConfig.get(defaultConfigEntry.getKey()) == null) {
+                            // entry exists in defaultConfig but doesn't exist in appConfig
+                            // so add it to result
+                            resultConfig.put(defaultConfigEntry.getKey(), defaultConfigEntry.getValue());
+                        }
+                    });
+            return resultConfig;
+        }
+    }
+
+    private static Map.Entry<String, Object>  configEntryMergeAndOverride(
+            Map.Entry<String, Object> appConfigEntry,
+            Map.Entry<String, Object> defaultConfigEntry) {
+
+        Object defaultConfigVal = defaultConfigEntry.getValue();
+
+        // base case
+        if (defaultConfigVal == null || defaultConfigVal instanceof String) {
```

Comment:
How about the case where `defaultConfigVal` was `instanceof` of `Map` but `appConfigEntry.getValue()` was instance of `string`? In this case, we'd want the `appConfigEntry.getValue()` to override.

Reply:
Good call! Corrected and simplified the conditional.